scipy/stats/_multivariate.py

```diff
@@ -2759,27 +2695,22 @@ def _logpdf(self, x, dim, df, scale, log_det_scale):
             Scale matrix
         log_det_scale : float
             Logarithm of the determinant of the scale matrix
+        C : ndarray
+            Cholesky factorization of the scale matrix, lower triagular.
 
         Notes
         -----
         As this function does no argument checking, it should not be
         called directly; use 'logpdf' instead.
 
         """
+        # Retrieve tr(scale x^{-1})
         log_det_x = np.empty(x.shape[-1])
-        x_inv = np.copy(x).T
-        if dim > 1:
-            _cho_inv_batch(x_inv)  # works in-place
-        else:
-            x_inv = 1./x_inv
         tr_scale_x_inv = np.empty(x.shape[-1])
-
         for i in range(x.shape[-1]):
-            C, lower = scipy.linalg.cho_factor(x[:, :, i], lower=True)
-
-            log_det_x[i] = 2 * np.sum(np.log(C.diagonal()))
-
-            tr_scale_x_inv[i] = np.dot(scale, x_inv[i]).trace()
+            Cx, log_det_x[i] = self._cholesky_logdet(x[:, :, i])
```

Comment:
Here since you don't do anything significant with A anyways hence that's just wasting memory, also the norm squared is just the trace hence `norm(A)` is quite expensive there. 

I think the old way is better with `cho_factor` but then we can add `cho_solve` in a more succint way, in other words

```python
Cx, lower = la.cho_factor(x[:, :, i], lower=True, check_finite=False)
log_det_x[i] = 2 * np.log(Cx.diagonal()).sum()
tr_scale_x_inv[i] = la.cho_solve((Cx, lower), C, check_finite=False).trace()
```

Reply:
The old version using `cho_solve` is effectively 2 calls to `trsm`, each of which is cubic in `dim`, followed by `trace`, which is linear in `dim`. The version I propose is a single call to `trtrs` (could be `trsm`), followed by `norm`, which is quadratic in `dim`, so I would expect it to be faster. I think the number of allocations are the same, so I don't see how the proposed version is more wasteful.

Reply:
Ah yes, my bad you are not solving the problem `scale @ X = C` but in fact solving the cholesky problem `Cx @ X = C`. Sorry about that I confused myself. 

```python
Cx, lower = la.cho_factor(x[:, :, i], lower=True, check_finite=False)
log_det_x[i] = 2 * np.log(Cx.diagonal()).sum()
tr_scale_x_inv[i] = la.blas.trsm(1., Cx, C, overwrite_b=True, ... etc. ).trace()
```

2-norm is basically svd which is O(n**3) and assumes a full matrix but you only have a triangular one. So regardless of what you do `norm` is quite expensive. Especially when the alternative is O(n).

Reply:
What I'm using here is the simplification
$$\mathrm{tr}(\Psi x^{-1}) = \langle \Psi, x^{-1} \rangle = \langle C C^T, C_X^{-T} C_X^{-1}  \rangle = \langle C_X^{-1} C, C_X^{-1} C  \rangle = \lVert C_X^{-1} C \rVert_2^2,$$
where $\lVert \cdot \rVert_2$ is the vector 2-norm (which only the lower triangle contributes to). The vector `norm` should be quite inexpensive compared to a second `trsm`, even if treating the matrix as dense.

```quote
2-norm is basically svd which is O(n**3) and assumes a full matrix but you only have a triangular one.
```
Are you referring to the matrix 2-norm? This is computing the vector 2-norm, which is O(n**2). According the docs for `norm`, only of we specify the `ord` keyword is the matrix norm computed.

```quote
So regardless of what you do `norm` is quite expensive. Especially when the alternative is O(n).
```
But the alternative is not `O(n)`. Since it includes 2 `trsm`s instead of 1, the alternative is `O(n**3)`. Besides, this implementation (modified from the most recent suggestion) will not work, since it computes $\mathrm{tr}(C_X^{-1} C)$, which is not the same as $\langle \Psi, x^{-1}\rangle$:
```python
tr_scale_x_inv[i] = scipy.linalg.blas.dtrsm(1., Cx, C, overwrite_b=True, lower=True, side=0).trace()
```

Reply:
in your version, in solve_triangular if Cx is matrix, and C is a matrix then A is also matrix. Your notation is getting a bit complicated here but then norm(A) should be a matrix norm then no? So I don't quite follow how it is vector norm.

Reply:
No, `norm(x)` for matrix `x` still computes a vector 2-norm unless `ord` is specified. From the docstring of `x` in [`norm`](https://numpy.org/devdocs/reference/generated/numpy.linalg.norm.html#numpy.linalg.norm):

```quote
If both _axis_ and _ord_ are None, the 2-norm of `x.ravel` will be returned.
```

A demo:
```python
>>> import numpy as np
>>> A = np.random.normal(size=(10, 10))
>>> np.linalg.norm(A.ravel())   # vector 2-norm
8.406980403795442
>>> np.dot(A.T, A).trace()**0.5 # vector 2-norm using inner product
8.406980403795444
>>> np.linalg.norm(A)           # when ord=None, vector norm
8.406980403795442
>>> np.linalg.norm(A, ord=2)    # when ord=2, matrix 2-norm
4.773833982984454
>>> np.linalg.svd(A)[1].max()   # explicit matrix 2-norm
4.7738339829844545
```

Reply:
OK I think we have a terminology issue. That's typically going with the name Frobenius norm but I guess you like the vector-2 name for it. But nevermind let's go with that definition, that was what I meant above. 

What I am saying is that, (again this notation is not so great, we can go with separate letters K, L, M instead of versions of $C$) You already have a cholesky factor $C$ and another cholesky factor $C_X$. You are then solving $C_X X = C$. Then computing the Frobenius norm of X (or vector 2-norm in your definition). So half of that entries are 0 anyways; X is also triangular but I guess I am splitting hair here so let's not dwell too much on it. I guess `n` will not be too much in practice anyways. 

Other than that, a separate comment I had was, replacing the `solve_triangular` with the naked `trsm` call and call norm on it because `solve_triangular` has a lot of overhead argument checking etc. But I guess things got convoluted sorry about that.

Reply:
No problem at all. I think we are on the same page now. I agree, notation is not great. It's not consistent in existing code though, so I've opted to stick with existing notation instead of creating a new one. (even in the `invwishart` docstring, it uses `\Psi` and `\Sigma` both to refer to the scale).

I benchmarked to make the choice of computing `norm(A)` instead of `norm(A[np.tril_indices(dim)])`. The first option seemed slightly faster, perhaps because the norm is sufficiently fast that the allocation is more expensive than computing the norm of the zero triangle. But I'll benchmark more carefully.

Sure, I've now switched to `trsm`.